File: src/main/java/pu/fmi/slavnarech/services/user/UserService.java
package pu.fmi.slavnarech.services.user;

import org.springframework.data.domain.Page;
import org.springframework.security.core.userdetails.UserDetailsService;
import pu.fmi.slavnarech.entities.connection.dtos.ConnectionResponse;
import pu.fmi.slavnarech.entities.member.MemberStatus;
import pu.fmi.slavnarech.entities.user.User;
import pu.fmi.slavnarech.entities.user.dtos.UserConnectionResponse;
import pu.fmi.slavnarech.entities.user.dtos.UserRequest;
import pu.fmi.slavnarech.entities.user.dtos.UserResponse;
import pu.fmi.slavnarech.utils.PageFilter;

public interface UserService extends UserDetailsService {

  Page<UserResponse> searchAllByUsername(PageFilter pageFilter);

  UserResponse saveUser(UserRequest userRequest);

  User registerUser(String username, String email, String password);

  User getById(Long id);

  User loadUserByEmail(String email);

  boolean changeStatusOfFriendRequest(Long id, Long connectionId, MemberStatus status);

  Page<UserConnectionResponse> getFriendsOfUser(Long id, PageFilter pageFilter);

  Page<ConnectionResponse> getChannelsOfUser(Long id, PageFilter pageFilter);

  Page<UserResponse> getFriendRequestsSendFromUser(Long id, PageFilter pageFilter);

  Page<UserConnectionResponse> getFriendInvitesReceivedForUser(Long id, PageFilter pageFilter);
}

File: src/main/frontend/src/app/services/user.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import {
  UserConnectionResponse,
  UserRequest,
  UserResponse,
} from '../models/user.models';
import { Observable } from 'rxjs';
import { Page, PageFilter } from '../models/page.modes';
import { MemberStatus } from '../enums/member-status.enum';
import { ConnectionResponse } from '../models/connection.modes';
import { MemberResponse } from '../models/member.models';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private httpClient = inject(HttpClient);
  private baseUrl = 'http://localhost:8080/api/v1/users';

  public createUser(userRequest: UserRequest) {
    return this.httpClient.post(this.baseUrl, userRequest);
  }

  public getAllUsers(
    params: HttpParams
  ): Observable<Page<UserConnectionResponse>> {
    return this.httpClient.get<Page<UserConnectionResponse>>(this.baseUrl, {
      params,
    });
  }

  public getFriendsOfUser(
    userId: number,
    params: HttpParams
  ): Observable<Page<UserConnectionResponse>> {
    return this.httpClient.get<Page<UserConnectionResponse>>(
      `${this.baseUrl}/${userId}/connections/friends`,
      {
        params,
      }
    );
  }

  public getChannelsOfUser(
    userId: number,
    params: HttpParams
  ): Observable<Page<ConnectionResponse>> {
    return this.httpClient.get<Page<ConnectionResponse>>(
      `${this.baseUrl}/${userId}/connections/channels`,
      {
        params,
      }
    );
  }

  public getFriendRequestsSendFromUser(
    userId: number,
    params: HttpParams
  ): Observable<Page<UserConnectionResponse>> {
    return this.httpClient.get<Page<UserConnectionResponse>>(
      `${this.baseUrl}/${userId}/connections/friend-requests`,
      {
        params,
      }
    );
  }

  public getFriendInvitesReceivedForUser(
    userId: number,
    params: HttpParams
  ): Observable<Page<UserConnectionResponse>> {
    return this.httpClient.get<Page<UserConnectionResponse>>(
      `${this.baseUrl}/${userId}/connections/friend-invites`,
      {
        params,
      }
    );
  }

  public changeStatusOfMemberFriend(
    userId: number,
    connectionId: number,
    memberStatus: MemberStatus
  ) {
    return this.httpClient.put(
      `${this.baseUrl}/${userId}/connection/${connectionId}/member/${memberStatus}`,
      null
    );
  }

  public getMemberByUserAndConnectionId(
    userId: number,
    connectionId: number
  ): Observable<MemberResponse> {
    return this.httpClient.get<MemberResponse>(
      `${this.baseUrl}/${userId}/connections/${connectionId}`
    );
  }
}
